### PR TITLE
SQL-2620: Add support for UNION distinct

### PR DESCRIPTION
**Pull request by bucaojit**
Implementing UNION distinct.  It differs a bit from the initial design, I applied some of the things we did for SELECT DISTINCT such as using the `__groupKey{}` alias.  
Added tests to confirm behavior, removed `skip_reasons` from the existing union distinct tests and moved them into `distinct.yml`

### CODE REVIEW

File: tests/spec_tests/query_tests/union.yml [hunk 1]
```diff
@@ -63,15 +44,6 @@ tests:
       - {'': {'a': 2}}
       - {'': {'a': 2}}
 
-  - description: union is left associative
-    current_db: mydb
-    query: "SELECT VALUE {'a': a} FROM foo UNION ALL SELECT VALUE {'a': a} FROM foo UNION SELECT VALUE {'a': a} FROM foo"
-    ordered: false
-    skip_reason: "distinct UNION not currently supported"
```
Comment (mattChiaravalloti) [inline tests/spec_tests/query_tests/union.yml:70]: Not sure why this one was skipped since it's doing a `UNION ALL`. Should we leave this one here? When you cut and pasted it into the other file, you left it as `UNION ALL` so it's out of place in the `distinct.yml` file.

File: tests/spec_tests/query_tests/distinct.yml
Comment (mattChiaravalloti) [inline tests/spec_tests/query_tests/distinct.yml:1]: Is `SELECT a, b FROM foo UNION SELECT a, b FROM bar` a valid query? If yes, can we also include a test for that demonstrating that the duplicate rows from the two result sets are deduplicated? We demonstrate this for selectin specific columns from the same collection (`SELECT a FROM foo UNION SELECT a FROM foo`) and we demonstrate this for same column names in a different order (`SELECT a, b FROM foo UNION SELECT b, a FROM bar`) and for `SELECT *` on both sides, but we don't demonstrate deduplication for same columns in the same order from different datasources on the two sides of a UNION. A test covering that would therefore be unique and useful.

File: mongosql/src/algebrizer/test.rs [hunk 2]
```diff
@@ -7454,20 +7454,256 @@ mod limit_or_offset_clause {
...
     test_algebrize!(
-        union_distinct_not_allowed,
+        union_distinct_star,
         method = algebrize_set_query,
-        expected = Err(Error::DistinctUnion),
-        expected_error_code = 3006,
+        expected = Ok(mir::Stage::Project(mir::Project {
+            source: Box::new(mir::Stage::Group(mir::Group {
+                source: Box::new(mir::Stage::Set(mir::Set {
+                    operation: mir::SetOperation::UnionAll,
```
Comment (mattChiaravalloti) [inline mongosql/src/algebrizer/test.rs:7474]: Ah very interesting revelation here: do we even need `mir::SetOperation::Union`? If we algebrize in such a way that all unions become `UnionAll`s in the `mir`, can we do away with `mir`'s representation of distinct `Union`?

File: mongosql/src/algebrizer/test.rs [hunk 3]
```diff
@@ -7454,20 +7454,256 @@ mod limit_or_offset_clause {
...
+                            db: "test".into(),
+                            collection: "bar".into(),
+                            cache: SchemaCache::new()
+                        })),
+                        expression: map! {
+                            (DatasourceName::Named("bar".into()), 0u16).into() => mir::Expression::Reference((DatasourceName::Named("bar".into()), 0u16).into())
+                        },
+                        is_add_fields: false,
+                        cache: SchemaCache::new()
+                    })),
```
Comment (mattChiaravalloti) [inline mongosql/src/algebrizer/test.rs:7498]:
```suggestion
                    left: Box::new(mir_source_foo()),
                    right: Box::new(mir_source_bar()),
```
We could reuse some of the global helpers here to reduce the size of these tests! We could even go a step further and simplify these two helpers into one that is parametrized by collection name. Up to you if you want to do that kind of refactoring here or not.

File: mongosql/src/algebrizer/test.rs [hunk 4]
```diff
@@ -7454,20 +7454,256 @@ mod limit_or_offset_clause {
...
+                    cache: SchemaCache::new()
+                })),
+                keys: vec![
+                    mir::OptionallyAliasedExpr::Aliased(mir::AliasedExpr {
+                        alias: "__groupKey0".to_string(),
+                        expr: mir::Expression::Reference(
+                            (DatasourceName::Named("bar".to_string()), 0u16).into()
+                        )
```
Comment (mattChiaravalloti) [inline mongosql/src/algebrizer/test.rs:7506]:
```suggestion
                        expr: mir::Expression::Reference(("bar", 0u16).into())
```
We have a `impl<T: Into<Key>> From<T> for ReferenceExpr` which utilizes `impl<D, S> From<(D, S)> for Key where D: Into<DatasourceName>, S: Into<u16>` which utilizes `impl<S> From<S> for DatasourceName where S: Into<String>`.

All of that to say, we ultimately allow converting `(&str, u16)` tuples into `ReferenceExpr`s directly! So you could remove the `DatasourceName::Named("...".to_string())` wherever you create an `Expression::Reference` to help cut down on the size of these tests!

File: mongosql/src/algebrizer/test.rs [hunk 5]
```diff
@@ -7454,20 +7454,256 @@ mod limit_or_offset_clause {
...
+                aggregations: vec![],
+                cache: SchemaCache::new(),
+                scope: 0
+            })),
+            expression: map! {
+                (DatasourceName::Named("bar".to_string()), 0u16).into() => mir::Expression::FieldAccess(mir::FieldAccess {
+                    expr: Box::new(mir::Expression::Reference((DatasourceName::Bottom, 0u16).into())),
+                    field: "__groupKey0".to_string(),
+                    is_nullable: true
+                }),
```
Comment (mattChiaravalloti) [inline mongosql/src/algebrizer/test.rs:7524]:
```suggestion
                ("bar", 0u16).into() => crate::util::mir_field_access("__bot__", "__groupKey0", true),
```
We have some more test helpers in `crate/util/mod.rs` that could again help reduce test size.

**Review by mattChiaravalloti — CHANGES_REQUESTED** [review]
This looks fantastic. The implementation is super slick! I left some comments directing you to some test helpers we have hidden away in a util file, as well as a couple suggestions for spec testing coverage and organization.

**Review by pmeredit — APPROVED** [review]
LGTM!

File: mongosql/src/algebrizer/test.rs (hunk 5 above)
Comment (pmeredit) [reply mongosql/src/algebrizer/test.rs:7524]: I think a good skunkworks project would be adding serde for mir so that we can make these tests less hideous :joy:

Comment (mattChiaravalloti) [reply mongosql/src/algebrizer/test.rs:7524]: That would be wonderful 🙏

File: tests/spec_tests/query_tests/union.yml (hunk 1 above)
Comment (bucaojit) [reply tests/spec_tests/query_tests/union.yml:70]: Hm, there's a UNION distinct on the right side to show that it is executing the `UNION ALL` on the left first to duplicate `foo` and then the UNION distinct will deduplicate it and so a:1 and a:2 only appears once.  Double checking if you missed the second UNION?  Otherwise sure I can move it back here in `union.yml`.

File: tests/spec_tests/query_tests/distinct.yml
Comment (bucaojit) [reply tests/spec_tests/query_tests/distinct.yml:1]: That's a good point, I had the mixed order test but not the same order `a,b`. I added a test to cover it.

File: mongosql/src/algebrizer/test.rs (hunk 4 above)
Comment (bucaojit) [reply mongosql/src/algebrizer/test.rs]: Sounds good.  I made this change and added a helper function to hopefully make things easier to follow.

File: mongosql/src/algebrizer/test.rs (hunk 3 above)
Comment (bucaojit) [reply mongosql/src/algebrizer/test.rs]: I added a parameterized helper function here.

**Review by mattChiaravalloti — APPROVED** [review]
